akka-sample-distributed-workers-scala/src/main/scala/worker/WorkManager.scala

```diff
@@ -19,169 +19,124 @@ import worker.WorkState.WorkerFailed
 
 /**
  * The work manager actor keep tracks of all available workers, and all scheduled and ongoing work items
  */
 object WorkManager {
 
+  val ManagerServiceKey = ServiceKey[ConsumerController.Command[WorkerCommand]]("worker-service-key")
+
   val WorkerServiceKey: ServiceKey[WorkerCommand] = ServiceKey[WorkerCommand]("workerService")
   val ResultsTopic = "results"
 
   final case class Ack(workId: String) extends CborSerializable
 
-  sealed trait WorkerStatus
-  case object Idle extends WorkerStatus
-  final case class Busy(workId: String, deadline: Deadline) extends WorkerStatus
-  final case class WorkerState(ref: ActorRef[WorkerCommand], status: WorkerStatus)
+//  sealed trait WorkerStatus
+//  case object Idle extends WorkerStatus
+//  final case class Busy(workId: String, deadline: Deadline) extends WorkerStatus
+//  final case class WorkerState(ref: ActorRef[WorkerCommand], status: WorkerStatus)
 
-  sealed trait Command extends CborSerializable
-  private final case class UpdatedWorkers(workers: Receptionist.Listing) extends Command
 
   // Messages from Workers
-  final case class WorkerRequestsWork(workerId: String, replyTo: ActorRef[WorkerCommand]) extends Command
-  final case class WorkIsDone(workerId: String, workId: String, result: Any, replyTo: ActorRef[WorkerCommand])
-      extends Command
-  final case class WorkFailed(worker: ActorRef[WorkerCommand], workId: String) extends Command
+ // final case class WorkerRequestsWork(workerId: String, replyTo: ActorRef[WorkerCommand]) extends Command
+  //final case class WorkIsDone(workerId: String, workId: String, result: Any, replyTo: ActorRef[WorkerCommand])
+  //    extends Command
+  //final case class WorkFailed(worker: ActorRef[WorkerCommand], workId: String) extends Command
 
   // Responses to requests from workers
   sealed trait WorkerCommand
   final case class DoWork(work: Work) extends WorkerCommand
-  final case class WorkAck(id: String) extends WorkerCommand
-  case object WorkAvailable extends WorkerCommand
 
-  // External commands
+  sealed trait Command extends CborSerializable
   final case class SubmitWork(work: Work, replyTo: ActorRef[WorkManager.Ack]) extends Command
+  private case class RequestNextWrapper(ask: RequestNext[WorkerCommand]) extends Command
+  final case class WorkIsDone(id: String) extends Command
+  final case class WorkFailed(id: String, t: Throwable) extends Command
+  private final case class TryStartWork() extends Command
 
   def apply(workTimeout: FiniteDuration): Behavior[Command] =
     Behaviors.setup { ctx =>
       Behaviors.withTimers { timers =>
         // FIXME use typed pub sub once https://github.com/akka/akka/issues/26338 is done
+
+        implicit val timeout = Timeout(5.seconds)
         val mediator = DistributedPubSub(ctx.system.toClassic).mediator
+        val producerController = ctx.spawn(WorkPullingProducerController[WorkerCommand]("work-manager", ManagerServiceKey, None), "producer-controller")
```

Comment:
I like the integration with the receptionist, i had done that my self intitially before realising